inference-engine/src/transformations/src/transformations/common_optimizations/mark_precision_sensitive_subgraphs.cpp

```diff
@@ -0,0 +1,73 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/common_optimizations/mark_precision_sensitive_subgraphs.hpp"
+
+#include <memory>
+#include <vector>
+
+#include "transformations/utils/utils.hpp"
+#include "transformations/rt_info/disable_fp16_compression.hpp"
+#include "openvino/opsets/opset1.hpp"
+#include "openvino/opsets/opset3.hpp"
+#include "openvino/opsets/opset8.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "openvino/op/util/precision_sensitive_attribute.hpp"
+
+using namespace std;
+
+namespace {
+void visit_shape_path(const shared_ptr<ov::Node>& node, set<shared_ptr<ov::Node>>& visited) {
+    if (!node)
+        return;
+    visited.insert(node);
+    deque<shared_ptr<ov::Node>> nodes{ node };
+    while (!nodes.empty()) {
+        auto curr_node = nodes.front();
+        nodes.pop_front();
+        // Do not check if already visited
+        if (ov::is_type<ov::opset1::ShapeOf>(curr_node) || ov::is_type<ov::opset3::ShapeOf>(curr_node)) {
+            continue;
+        }
+        visited.insert(curr_node);
+        if (ov::is_type<ov::opset8::Constant>(curr_node)) {
+            ov::disable_fp16_compression(curr_node);
+        } else {
+            for (auto& input_value : curr_node->input_values()) {
+                // continue searching
+                const auto& input_node = input_value.get_node_shared_ptr();
+                nodes.push_front(input_node);
```

Comment:
not clear how you handle ShapeOf subgraph from outputs to inputs? could you please clarify?

Reply:
This transformation goes from precision-sensitive inputs to the ShapeOf nodes and marks such subgraph as disabled for FP16 conversion.